TrkReco/src/RobustHelixFit.cc

```diff
@@ -696,15 +697,251 @@ namespace mu2e
+	  // }
+	  swmax[ipeak] = sw;
+	}
+      }//end loop pver the bins of the array 'hist'
+
+      if (swmax[ipeak] <= 0)       break;
+      float numerator_sum(0);
+    
+      //calculating the standard deviation of the peak
+      for (int l=0; l<binsToIntegrate; ++l){
+	sw             = hist_sum[indexPeak[ipeak]+l]; //hdz->GetBinContent(indexPeak[ipeak]+l+1);
+	float        x = (indexPeak[ipeak] + l + .5)*bin_size +  start_dz;  //hdz->GetBinCenter(indexPeak[ipeak] + l+1);
+	numerator_sum += sw*(x - xmp[ipeak])*(x - xmp[ipeak]);
+      }
+      sigma[ipeak] = sqrt((numerator_sum)/swmax[ipeak]);
+      
+      //checking if peak postion is far enogh away from dz=0
+      if (xmp[ipeak]-nsigma*sigma[ipeak]>0) {
+	if (peaks_found == 0) first_peak = ipeak;
+	peaks_found = peaks_found +1;
+      }
+
+    }//end the loop over npeaks
+
+    float     total_wg(0), weight_lambda(0);
+    for (int i=1; i<nMaxPeaks; ++i){
+      if (swmax[i]>minNCounts){
+	if ( (xmp[i]-nsigma*sigma[i]>0) && (xmp[i-1]-nsigma*sigma[i-1]>0)){
+	  double   wg = sqrt(swmax[i]*swmax[i-1]);
+	  weight_lambda += wg*(xmp[i] - xmp[i-1])/6.28;//(lambda[i]*swmax[i]);
+	  total_wg   += wg;
+	}
+      }
+    }
+    weight_lambda /= total_wg;
+
+    if (peaks_found == 1) weight_lambda = xmp[first_peak]/6.28;
+
+    float lambda_final = weight_lambda*dzdphisign;// extract_result<tag::weighted_median>(accf);
+
+    if(!goodLambda( rhel.helicity(),lambda_final) ) return retval;
+    rhel._lambda = lambda_final;
+
+    if (_diag){
+      HelixData._diag.lambda_0 = rhel._lambda;
+    }
+
+    // find phi at z intercept.  Use a histogram technique since phi looping
+    // hasn't been resolved yet, and to avoid inefficiency at the phi wrapping edge
+    _hphi.Reset();
+    for (int f=0; f<nHits; ++f){
+      hitP1 = &HelixData._chHitsToProcess[f];
+      if (!use(*hitP1) )             continue;   
+      
+      float phiex = rhel.circleAzimuth(hitP1->pos().z());
+      float dphi  = deltaPhi(phiex,hitP1->helixPhi());
+      _hphi.Fill(dphi);
+      _hphi.Fill(dphi-CLHEP::twopi);
+      _hphi.Fill(dphi+CLHEP::twopi);
+    }//end loop over the hits
+
+
+    // take the average of the maximum bin +- 1
+    int imax = _hphi.GetMaximumBin();
+    // if ( (imax <1) || (imax>_nphibins)) ###put excpetion throw
+    float count(0.0);
+    float fz0(0.0);
+    for (int ibin=std::max((int)0,imax-1); ibin <= std::min((int)imax+1,(int)_nphibins); ++ibin)
+      {
+	count += _hphi.GetBinContent(ibin);
+	fz0   += _hphi.GetBinContent(ibin)*_hphi.GetBinCenter(ibin);
+      }
+    
+    if (_diag){
+      HelixData._diag.nfz0counter = count;
+    }
+
+    if(count > _minnphi)
+      {
+	fz0 /= count;
+	rhel._fz0 = deltaPhi(0.0,fz0);
+
+	for (int f=0; f<nHits; ++f){
+	  hitP1 = &HelixData._chHitsToProcess[f];
+	  // if (!use(*hitP1) )                          continue;   
+	  resolvePhi(*hitP1,rhel);
+	}
+	
+	retval = true;
+      }
+
+    return retval;
```

Comment:
This function is >200 lines long, which makes it hard to read and maintain.  I think it logically splits into several sub-functions: fill the histogram, find a peak, average around the peak, find intercept, etc.  The sub-functions might be useful in other contexts too.